core/isolate.rs

```diff
@@ -36,6 +36,7 @@ pub type OpAsyncFuture<E> = Box<dyn Future<Item = Buf, Error = E> + Send>;
 pub enum Op<E> {
   Sync(Buf),
   Async(OpAsyncFuture<E>),
+  AsyncOptional(OpAsyncFuture<E>),
```

Comment:
Hmm - what is the purpose of this? I'm not open to adding to this enum...

Reply:
It is a marker such that when later registering the op we can tell if the op is optional and should not block event loop exit (by later incrementing optional_ops_count on `Isolate`). It is used since I don't think there is some easier way to pass this bit of information down other than with a new enum (otherwise we need to change the signature of `Async`)

Reply:
The design of signal is by first calling `SignalListen` to add a signal stream to resources. Then signal is actually handled by repeatedly calling `SignalPoll` to wait for next poll of signal -- somewhat similar to what we do with `TcpStream` read. But unlike reading from `TcpStream`, the task registered by `SignalPoll` is optional, such that if all other required tasks has finished (pending_ops only contains optional ops), program would exit right away.